### PR TITLE
Repairbots

### PR DESCRIPTION
**Pull request by Ben10Omintrix**
## About The Pull Request
adds repairbots to the game!
repairbot

this pr serves as a massive rework and buff to floorbots. i was a bit sad that they dont get built much anymore so ive given them tons of more utilities and uses.
Repairbots still inherit to place tiles and repair breaches. but they can now rebuild walls, rebuild windows and repair structure and machinery. Also Ive given them voicelines to add more character to them. In short, they are very depressed with their job (however they express their happy go lucky attitude when u emag them where they will start deconstructing the station)

to demonstrate capabilities, here's a slightly sped up clip of some repairbots patching up an area that was maxcapped 4 times:
https://github.com/user-attachments/assets/bddac3b0-1984-4571-85d3-c5283dd7c0de

When repairbots feel threatened, they will retract into their little toolbox which u can pick up and hold in ur hand, either to conveniently carry and plop them down at breached sites, or to bash skull with it

U can build them using a toolbox, proximity sensor, cyborg arm and a conveyor belt

## Why It's Good For The Game
refactors floorbots and makes them alot more useful tools for engineers to use

## Changelog
:cl:
refactor: floorbots have been refactored, please report any bugs
add: adds repairbots to the game!
/:cl:

### CODE REVIEW

**Comment by DrAmazing343** [comment]
yet another Ben10O slam-dunk. god bless...

**Comment by ZephyrTFA** [comment]
How are these different to drones

**Comment by Ben10Omintrix** [comment]
> How are these different to drones

main difference between these against drones and engiborgs is that these are only AI controlled whereas the others depend on player input. other than that theres some upsides and downsides to building these instead of the other types i mentioned. drones/engiborgs are whole lot surperior in that they can interact with more things but a repairbot is going to offer more consistency with the things it can build since itll automatically patrol the station at all times. theyre also very configurable since u can program which tasks they focus on through their interface. i also plan to make them able to respond to engineering emergencies sent out on the announcements system, whereas drones would be limited in that since they dont understand common

**Comment by AMyriad** [comment]
RIP no more recycling toolboxes
very impressive stuff though, I had to pick up my jaw after watching that video

**Comment by theselfish** [comment]
image

**Comment by Jacquerel** [comment]
My objection to this will sound insane but I think these are too cute.
Our existing bots are very clearly half-baked and cobbled together out of objects, whereas this one _has a face_.
Personally I much prefer bots to be like... depersonalised objects that gain their cuteness through action rather than because they're visually appearing.

**Comment by alien3301** [comment]
They look overpowered as hell not gonna lie. Engineers and drones already don't get to fix things much but with a roboticist shitting out 2 of those guys per minute repairs will basically never be needed aside from piping and wiring.

I think the craft should use an RCD since they use the RCD animation and RCDs are a good way to limit the amount of them on a station.

**Comment by the-orange-cow** [comment]
These look great, love the sprites!

**Comment by Nerev4r** [comment]
i'm chill w/ them being strong/useful but consider making them just require a source of mats; maybe they hit up and ruthlessly tusken raid the nearest autolathe?

**Comment by Wallemations** [comment]
2 things
1) What do they do when they run out of materials to build with? Do they return to where they were built? Do they just go back to patrolling and ignore everything? Do they say anything over the radio?
2) Do they also say their voice lines in-game as a readable statement, or are those just sound files only?

**Comment by Arturlang** [comment]
Would be nice if the building animations wouldn't loop multiple times

**Comment by Exester509** [comment]
> My objection to this will sound insane but I think these are too cute. Our existing bots are very clearly half-baked and cobbled together out of objects, whereas this one _has a face_. Personally I much prefer bots to be like... depersonalised objects that gain their cuteness through action rather than because they're visually appearing.
> 
> I really _would_ rather have a toolbox with wheels and a blinking light than something which is recognisably a little guy.

That's a fair assessment, the current lineup of bots gives off feels of extreme utilitarianism, this ultimately looks more similar to a mini-cyborg of sorts than to the rest of the bots

**Comment by Ben10Omintrix** [comment]
> To confirm, you mentioned they can rebuild windows, but it looks like the video you linked is meta arrivals and it looks like they just walled the entire thing in. Do they try to return the room they're working in to roundstart, or do they just try to close the area in to close breaches? If the latter, do they fight with people who tear down walls to expand rooms? Will they wall off the shuttle exits in evac in silence after meteors breach the whole thing?
> 
> With all that said, I'm all for this

They'll repair cracked windows, and if there's window frames without any glass they'll place windows over them. they shouldnt rebuild any walls that anyone tears down as long as it doesnt lead to space. and yes they would hypothetically wall off shuttle exits but im still looking for a nice solution to that

**Comment by Ben10Omintrix** [comment]
> i'm chill w/ them being strong/useful but consider making them just require a source of mats; maybe they hit up and ruthlessly tusken raid the nearest autolathe?

they do need people to insert mats into them before they do anything. i just had them all start out with a ton of them for the purpose of the demo

**Comment by Ben10Omintrix** [comment]
> 2 things
> 
> 1. What do they do when they run out of materials to build with? Do they return to where they were built? Do they just go back to patrolling and ignore everything? Do they say anything over the radio?
> 2. Do they also say their voice lines in-game as a readable statement, or are those just sound files only?

Currently they dont do anything when ran out of mats but i like ur idea. i could have them return to the roboticist and politely ask for a refill.   their voicelines are both readable and hearable

File: _maps/map_files/NorthStar/north_star.dmm [hunk 1]
```diff
@@ -36553,7 +36553,7 @@
 "jBf" = (
 /obj/effect/decal/cleanable/dirt,
 /obj/structure/table/reinforced,
-/obj/item/bot_assembly/floorbot,
+/ovj/item/bot_assembly/repairbot,
```
Comment (Sadboysuss) [inline _maps/map_files/NorthStar/north_star.dmm:36556]:
```suggestion
/obj/item/bot_assembly/repairbot,
```

**Comment by Livrah** [comment]
Can it use custom tiles as old floorbots?

**Comment by TheVekter** [comment]
Absolute fire, only complaint I have is that the end product doesn't _look_ like it's made with a toolbox.

**Comment by ZephyrTFA** [comment]
> > 2 things
> > 
> > 1. What do they do when they run out of materials to build with? Do they return to where they were built? Do they just go back to patrolling and ignore everything? Do they say anything over the radio?
> > 2. Do they also say their voice lines in-game as a readable statement, or are those just sound files only?
> 
> Currently they dont do anything when ran out of mats but i like ur idea. i could have them return to the roboticist and politely ask for a refill.   their voicelines are both readable and hearable

Please give them the ability to mug engineering borgs.

**Comment by PeriodicChaos** [comment]
> > 2 things
> > 
> > 1. What do they do when they run out of materials to build with? Do they return to where they were built? Do they just go back to patrolling and ignore everything? Do they say anything over the radio?
> > 2. Do they also say their voice lines in-game as a readable statement, or are those just sound files only?
> 
> Currently they dont do anything when ran out of mats but i like ur idea. i could have them return to the roboticist and politely ask for a refill. their voicelines are both readable and hearable

Could you make them take mats on the floor or tables? When areas get blown up the walls and stuff usually drop mats it would be cool if they could recycle it. Also could help out an ai who has no borgos that way. 👍

**Comment by Admiral-Gigaton** [comment]
Can you, and this might be stretch, but make them able to lay cable? That way whrn they place tiles all over, engies dont have to go tear it all up and lay cabling, but still need to reinstall APCs, air alarms, and all that


Factorio construction bot swarm is becoming real, just need to automate that shit

**Comment by WarerBrow** [comment]
Extremely overpowered. Maxcaps do not mean a thing now. 
But extremely useful. Make this a late-game tech, which wouldn't _replace_ floorbots, but instead it'd be an upgraded version of them. Also make them very **expensive** to build. RCD + bluespace/diamonds.

**Comment by TheVekter** [comment]
> Extremely overpowered. Maxcaps do not mean a thing now. But extremely useful. Make this a late-game tech, which wouldn't _replace_ floorbots, but instead it'd be an upgraded version of them. Also make them very **expensive** to build. RCD + bluespace/diamonds (or a charged anomaly core)

Clown take tbh, maxcaps do something which is "blow the fuck out of anyone inside them", plus these bots are relatively slow. If you watch the sped up video that Ben posted, like ten of them take quite a while to fix a multi maxcap sized hole. If you're only using bombs to try and get the crew to throw their hands up and just leave, use more.

**Comment by Ben10Omintrix** [comment]
> Absolute fire, only complaint I have is that the end product doesn't _look_ like it's made with a toolbox.

I may have to end up tweaking its design to fulfil jacquerel's request. If I can't come up with a decent design tho I'll likely just revert to the old floorbot sprite

**Comment by Ben10Omintrix** [comment]
> > > 2 things
> > > 
> > > 1. What do they do when they run out of materials to build with? Do they return to where they were built? Do they just go back to patrolling and ignore everything? Do they say anything over the radio?
> > > 2. Do they also say their voice lines in-game as a readable statement, or are those just sound files only?
> > 
> > Currently they dont do anything when ran out of mats but i like ur idea. i could have them return to the roboticist and politely ask for a refill.   their voicelines are both readable and hearable
> 
> Please give them the ability to mug engineering borgs.

Consider it done they WILL be robbing people in maintenance alleyways

**Comment by Ben10Omintrix** [comment]
> > > 2 things
> 
> > > 
> 
> > > 1. What do they do when they run out of materials to build with? Do they return to where they were built? Do they just go back to patrolling and ignore everything? Do they say anything over the radio?
> 
> > > 2. Do they also say their voice lines in-game as a readable statement, or are those just sound files only?
> 
> > 
> 
> > Currently they dont do anything when ran out of mats but i like ur idea. i could have them return to the roboticist and politely ask for a refill. their voicelines are both readable and hearable
> 
> 
> 
> Could you make them take mats on the floor or tables? When areas get blown up the walls and stuff usually drop mats it would be cool if they could recycle it. Also could help out an ai who has no borgos that way. 👍

good idea

**Comment by Ben10Omintrix** [comment]
> Can you, and this might be stretch, but make them able to lay cable? That way whrn they place tiles all over, engies dont have to go tear it all up and lay cabling, but still need to reinstall APCs, air alarms, and all that
> 
> 
> 
> 
> 
> Factorio construction bot swarm is becoming real, just need to automate that shit

It'll be a bit complicated to code this behavior but I could definitely do it in a future PR

**Comment by Exester509** [comment]
Actually, rather important question that just popped up in my head.
What do they do when they're emagged? Do they run around deconstructing stuff? Search for materials and get rid of them? Build walls EVERYWHERE?

**Comment by Dorsisdwarf** [comment]
The PR body states that when emagged they begin deconstructing the station, but I can't seem to find any ai behaviour for this, or actually any emag/hack interaction beyond more menacing voicelines. Am I misunderstanding the behaviour tree?

File: code/modules/mob/living/basic/bots/bot_hud.dm [hunk 2]
```diff
@@ -27,7 +27,7 @@
 	switch(mode)
 		if(BOT_SUMMON, BOT_RESPONDING) //Responding to PDA or AI summons
 			holder.icon_state = "hudcalled"
-		if(BOT_CLEANING, BOT_REPAIRING, BOT_HEALING) //Cleanbot cleaning, Floorbot fixing, or Medibot Healing
+		if(BOT_CLEANING, BOT_HEALING) //Cleanbot cleaning, Floorbot fixing, or Medibot Healing
```
Comment (Dorsisdwarf) [inline code/modules/mob/living/basic/bots/bot_hud.dm:30]: Comment reference to floorbots still left in here

**Comment by Ben10Omintrix** [comment]
> Actually, rather important question that just popped up in my head. What do they do when they're emagged? Do they run around deconstructing stuff? Search for materials and get rid of them? Build walls EVERYWHERE?

they will start deconstructing things

**Comment by LSommers** [comment]
A possible solution for the previously mentioned "what about the cables" and "what about engineering's job" might be to have them not replace floor tiles at all- just the plating. This leaves missing cables and pipes more visible, and also encourages engineering to fix them- or some random assistant to put tile down and leave the engineers cursing all greyshirts as they pry them up to fix things later.

**Review by Wallemations — CHANGES_REQUESTED** [review]
I'm a bit torn on the sprites
On one hand, I think it's really cute!
On the other hand, I feel like the fact that it's shoved into a toolbox has been kinda diminished? Idk if there's any way to increase that?

**Comment by Ben10Omintrix** [comment]
> I'm a bit torn on the sprites On one hand, I think it's really cute! On the other hand, I feel like the fact that it's shoved into a toolbox has been kinda diminished? Idk if there's any way to increase that?

im open to suggestions if u have any. i can make the toolbox a little bit bigger and more apparent

**Comment by Wallemations** [comment]
> im open to suggestions if u have any. i can make the toolbox a little bit bigger and more apparent

I'd have to really think on how to incorporate the toolbox better tbh
My suggest would have been maybe make the toolbox part wider but then the arms would be messed with and the arms are kinda important to keep for your new design. We don't want the arms to be poking out of the top of a toolbox, that'd make it look a bit goofy.
Maybe you could ask in the sprite channel in the discord for some more suggestions?

**Comment by nikothedude** [comment]
What if the repairbot sought out carbons with damaged prosthetic limbs, making them a sort of synthetic medbot? I think that'd fit in very well with their theme.

**Comment by deathrobotpunch** [comment]
> What if the repairbot sought out carbons with damaged prosthetic limbs, making them a sort of synthetic medbot? I think that'd fit in very well with their theme.

if that’s all they did then they would never be used

**Comment by Ben10Omintrix** [comment]
> What if the repairbot sought out carbons with damaged prosthetic limbs, making them a sort of synthetic medbot? I think that'd fit in very well with their theme.

i really wouldve. but the AI is already getting a bit out of hand (approaching 400 lines). i can do it down the line in a future pr

**Comment by Wallemations** [comment]
> > What if the repairbot sought out carbons with damaged prosthetic limbs, making them a sort of synthetic medbot? I think that'd fit in very well with their theme.
> 
> 
> 
> i really wouldve. but the AI is already getting a bit out of hand (approaching 400 lines). i can do it down the line in a future pr

I feel like that kinda negates part of the downside of choosing prosthetic limbs; that you need to go out of your way to heal them.
Not to mention I feel it'll make the repair bot even more OP, since let's be honest, being able to send a little guy out to fix an entire explosion is way more powerful than just fixing floors. If they also could heal players I think that'd be going a bit overboard.

**Comment by Ben10Omintrix** [comment]
codewise this is ready, i just have to figure out what im gonna do with the sprite

**Review by Wallemations — APPROVED** [review]
Looking at it again after giving it a lot of thought I think the sprite is fine enough where I won't hold the pr back because of it.

**Comment by TheVekter** [comment]
I really think using the old sprite is fine. I feel like the bots looking like they're cobbled together from what was used to make them is really important and I don't love that the new version looks so futuristic.

**Comment by Ghommie** [comment]
Ping me once CI passes and I'll review it (probably).

**Comment by Ben10Omintrix** [comment]
Should probably be TMed for a bit

File: code/__DEFINES/robots.dm [hunk 3]
```diff
@@ -364,3 +362,16 @@ DEFINE_BITFIELD(janitor_mode_flags, list(
 #define MEDIBOT_VOICED_THE_END "Is this the end?"
 #define MEDIBOT_VOICED_NOOO	"Nooo!"
 #define MEDIBOT_VOICED_CHICKEN "LOOK AT ME?! I am a chicken."
+
+//repairbot neutral voicelines
+#define REPAIRBOT_VOICED_HOLE "patching holes... but who is going to patch the hole in my heart..."
```
Comment (Ghommie) [inline code/__DEFINES/robots.dm:367]: I don't think these have to be defined, but whatever.

File: code/modules/mob/living/basic/bots/bot_hud.dm (hunk 2 above)
Comment (Ghommie) [inline code/modules/mob/living/basic/bots/bot_hud.dm:30]: This still references floorbots.

**Review by Ghommie — APPROVED** [review]

**Comment by LemonInTheDark** [comment]
massive soul death tgstation has fallen

**Comment by Chowder-McArthor** [comment]
Repairbot crafting;
Manual crafting needs the conveyor belt, and will use up a whole stack instead of just 1.
Crafting menu crafting still asks for 10 floor tiles.

**Comment by LemonInTheDark** [comment]
is this an issue report? if so this is the wrong place for it please use <https://github.com/tgstation/tgstation/issues/new?assignees=&labels=&projects=&template=bug_report_form.yml>

**Comment by Chowder-McArthor** [comment]
Ah sorry, just figured this would've been faster.

**Comment by LemonInTheDark** [comment]
sall good just a matter of categorization

**Comment by Ben10Omintrix** [comment]
> Ah sorry, just figured this would've been faster.  There's actually a number of problems with this PR I'm having.

Feel free to message me about them on the discord as well. My username's Kobsa on there

**Pull request merged**